In [1]:
import pandas as pd


In [2]:
mls = pd.read_csv('usa (1).csv')

In [3]:
mls.head()

Country League  Season        Date   Time                  Home  \
0     USA    MLS    2012  10/03/2012  23:00       Colorado Rapids   
1     USA    MLS    2012  10/03/2012  23:00   Vancouver Whitecaps   
2     USA    MLS    2012  11/03/2012  00:30             DC United   
3     USA    MLS    2012  11/03/2012  03:30    Los Angeles Galaxy   
4     USA    MLS    2012  11/03/2012  03:30  San Jose Earthquakes   

                     Away  HG  AG Res    PH    PD    PA  MaxH  MaxD  MaxA  \
0           Columbus Crew   2   0   H  2.06  3.40  4.11  2.13  3.40  4.11   
1         Montreal Impact   2   0   H  2.04  3.44  4.11  2.04  3.44  4.11   
2    Sporting Kansas City   0   1   A  2.57  3.31  3.04  2.60  3.60  3.04   
3          Real Salt Lake   1   3   A  1.88  3.57  4.88  1.89  3.57  5.00   
4  New England Revolution   1   0   H  1.83  3.65  4.95  1.91  3.68  5.00   

   AvgH  AvgD  AvgA  
0  2.01  3.19  3.65  
1  1.97  3.27  3.70  
2  2.46  3.11  2.81  
3  1.81  3.33  4.31  
4  1.82  3.36  4.23

In [4]:
mls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2948 entries, 0 to 2947
Data columns (total 19 columns):
Country    2948 non-null object
League     2948 non-null object
Season     2948 non-null int64
Date       2948 non-null object
Time       2948 non-null object
Home       2948 non-null object
Away       2948 non-null object
HG         2948 non-null int64
AG         2948 non-null int64
Res        2948 non-null object
PH         2948 non-null float64
PD         2948 non-null float64
PA         2948 non-null float64
MaxH       2948 non-null float64
MaxD       2948 non-null float64
MaxA       2948 non-null float64
AvgH       2948 non-null float64
AvgD       2948 non-null float64
AvgA       2948 non-null float64
dtypes: float64(9), int64(3), object(7)
memory usage: 437.7+ KB


date column is a object.  should i convert to datatype: datetime?

df['A'] = pd.to_datetime(df['A'])

drop some unnecessary columns related to sports betting odds

In [5]:
columns = ['Country', 'PH', 'PD', 'PA', 'MaxH', 'MaxD', 'MaxA', 'AvgH', 'AvgD', 'AvgA']
mls.drop(columns, inplace=True, axis=1)

list of all unique mls teams, then map out their cities

first get a list of all unique values in the column using : #List unique values in the df['name'] column
df.name.unique() as a template

In [6]:
cities = mls.Home.unique()

In [7]:
print(cities)

['Colorado Rapids' 'Vancouver Whitecaps' 'DC United' 'Los Angeles Galaxy'
 'San Jose Earthquakes' 'FC Dallas' 'Chivas USA' 'Portland Timbers'
 'Montreal Impact' 'Sporting Kansas City' 'Real Salt Lake'
 'Seattle Sounders' 'Philadelphia Union' 'Toronto FC' 'Columbus Crew'
 'New England Revolution' 'Chicago Fire' 'New York Red Bulls'
 'Houston Dynamo' 'Orlando City' 'New York City' 'Atlanta United'
 'Minnesota United' 'Los Angeles FC' 'FC Cincinnati' 'Atlanta Utd']


noteworthy data quality:  Atlanta duplicated as United and Utd

Now, we want to replace values in a column:  home team to city using this as a template: w['female'] = w['female'].map({'female': 1, 'male': 0})

In [8]:
mls["City"] = mls["Home"].map({'Atlanta Utd': 'Atlanta, GA', 'FC Cincinnati': 'Cincinnati, OH', 'Los Angeles FC': 'Los Angeles, CA', 'Minnesota United': 'Minneapolis, MN', 'Atlanta United': 'Atlanta, GA', 'New York City':'New York, NY', 'Orlando City': 'Orlando, FL', 'Houston Dynamo': 'Houston, TX', 'New York Red Bulls': 'New York, NY', 'Chicago Fire': 'Chicago, IL', 'New England Revolution': 'Boston, MA', 'Columbus Crew': 'Columbus, OH', 'Toronto FC': 'Toronto, Ontario', 'Philadelphia Union': 'Philadelphia, PA', 'Seattle Sounders': 'Seattle, WA', 'Real Salt Lake':'Salt Lake City, UT', 'Sporting Kansas City':'Kansas City, MO', 'Montreal Impact' : 'Montreal, Quebec', 'Chivas USA': 'Los Angeles, CA' , 'FC Dallas': 'Dallas, TX', 'Colorado Rapids': 'Denver, CO', 'Vancouver Whitecaps': 'Vancouver, Canada', 'DC United': 'Washington, DC', 'Los Angeles Galaxy': 'Los Angeles, CA', 'San Jose Earthquakes': 'San Jose, CA' })

In [9]:
mls.head()

League  Season        Date   Time                  Home  \
0    MLS    2012  10/03/2012  23:00       Colorado Rapids   
1    MLS    2012  10/03/2012  23:00   Vancouver Whitecaps   
2    MLS    2012  11/03/2012  00:30             DC United   
3    MLS    2012  11/03/2012  03:30    Los Angeles Galaxy   
4    MLS    2012  11/03/2012  03:30  San Jose Earthquakes   

                     Away  HG  AG Res               City  
0           Columbus Crew   2   0   H         Denver, CO  
1         Montreal Impact   2   0   H  Vancouver, Canada  
2    Sporting Kansas City   0   1   A     Washington, DC  
3          Real Salt Lake   1   3   A    Los Angeles, CA  
4  New England Revolution   1   0   H       San Jose, CA

draws = mls.loc[mls['Res'] == 'D']

719 draws out of 2948 total matches

In [21]:
list_of_dates_with_draws = mls.loc[mls['Res'] == 'D']

In [24]:
print(list_of_dates_with_draws.tail())

     League  Season        Date   Time                  Home  \
2922    MLS    2019  30/09/2019  00:30  Sporting Kansas City   
2924    MLS    2019  06/10/2019  21:00             DC United   
2936    MLS    2019  19/10/2019  20:30      Seattle Sounders   
2937    MLS    2019  19/10/2019  23:00            Toronto FC   
2939    MLS    2019  20/10/2019  20:00    Philadelphia Union   

                    Away  HG  AG Res              City  
2922    Portland Timbers   2   2   D   Kansas City, MO  
2924       FC Cincinnati   0   0   D    Washington, DC  
2936           FC Dallas   3   3   D       Seattle, WA  
2937           DC United   1   1   D  Toronto, Ontario  
2939  New York Red Bulls   3   3   D  Philadelphia, PA  


In [12]:
print(len(list_of_dates_with_draws))

719


In [13]:
list_of_times_of_draws = mls.loc[mls['Res'] == 'D']['Time'].tolist()

In [14]:
list_of_cities_of_draws = mls.loc[mls['Res'] == 'D']['City'].tolist()

Do Python lists maintain order?
Yes, the order of elements in a python list is persistent. In short, yes, the order is preserved. via stackoverflow Dec 4, 2012

do i need to convert the time to local time?  

In [18]:
print(list_of_cities_of_draws)

['Montreal, Quebec', 'Dallas, TX', 'Vancouver, Canada', 'Philadelphia, PA', 'Washington, DC', 'New York, NY', 'Chicago, IL', 'Washington, DC', 'Columbus, OH', 'Dallas, TX', 'Los Angeles, CA', nan, 'Chicago, IL', 'Montreal, Quebec', 'San Jose, CA', 'Houston, TX', 'Vancouver, Canada', 'Boston, MA', 'Dallas, TX', 'Denver, CO', 'San Jose, CA', 'New York, NY', nan, 'Los Angeles, CA', 'Boston, MA', 'Houston, TX', 'Seattle, WA', 'Vancouver, Canada', 'Toronto, Ontario', 'Dallas, TX', 'Toronto, Ontario', 'Boston, MA', 'Houston, TX', 'Dallas, TX', 'Salt Lake City, UT', 'Kansas City, MO', 'Dallas, TX', 'Los Angeles, CA', 'New York, NY', 'Vancouver, Canada', 'Kansas City, MO', 'San Jose, CA', nan, 'Columbus, OH', 'Toronto, Ontario', 'Denver, CO', 'Washington, DC', 'Houston, TX', 'Philadelphia, PA', 'Houston, TX', 'Kansas City, MO', 'Washington, DC', 'Boston, MA', 'Boston, MA', 'Dallas, TX', 'Kansas City, MO', 'Toronto, Ontario', nan, 'San Jose, CA', 'Montreal, Quebec', 'Boston, MA', 'Vancouver, Ca